Base time API upon std::chrono

251 opened the pull request:
This should not change the behaviour of KLEE and mimics the old API.

- functions moved from util into time namespace
- uses time points and time spans instead of "context-sensitive" double
- CLI time arguments now can have the form "3h5min8us"

Changed command line parameters:
- batch-time (double to string)
- istats-write-interval (double to string)
- max-instruction-time (double to string)
- max-solver-time (double to string)
- max-time (double to string)
- min-query-time-to-log (double to string)
- seed-time (double to string)
- stats-write-interval (double to string)
- uncovered-update-interval (double to string)
- added: log-timed-out-queries (replaces negative max-solver-time, default: on)

This PR closes #210, #207 and avoids ifdef hell as in #674.

I measured the performance before implementing that interface. In short:
- gettime/coarse: adv: fastest, disadv: software clock (resolution jiffies), linux-specific
- gettime/monotonic: adv: POSIX, fastest on Linux; disadv: slowest on OS X (by far)
- chrono/steady: adv: C++11, almost as fast as gettime/monotonic, implemented in this PR
- gettimeofday: adv: slightly faster than std::chrono on OS X, disadv: C-like API

I've introduced `time::Point` and `time::Span` classes as C++ got the defaults wrong as always (e.g. Duration is not 0 initialised).

kren1 commented on include/klee/Internal/System/Time.h:
```diff
@@ -10,20 +10,112 @@
+    using SteadyTimePoint = std::chrono::steady_clock::time_point;
+    using Duration = std::chrono::steady_clock::duration;
+
+    struct Point;
```
I don't think these need to be forward declared.

kren1 commented on include/klee/Internal/System/Time.h:
```diff
@@ -10,20 +10,112 @@
+    Point getWallTime();
+
+    /// Parses a string to span
+    Span spanFromString(const std::string&);
```
Why is this not just a `Span` constructor?

251 replied:
Because it doesn't need access to members:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-member

kren1 replied:
But you already have a function that is a member (the constructor), so I don't think this applies.

251 replied:
I see, makes sense.

kren1 commented on lib/Basic/CmdLineOptions.cpp:
```diff
@@ -47,19 +47,19 @@ cl::opt<bool>
+cl::opt<bool>
+LogTimedOutQueries("log-timed-out-queries",
+                   cl::init(true),
+                   cl::desc("Log timed out queries. (default=true)."));
```
This should be default=false, to preserve current behaviour?

251 replied:
Yeah, I had it that way first. But it should happen rarely and when it happens I typically want to know why without rerunning everything. But sure, I can change that.

kren1 replied:
I guess if it's a separate option it might make sense to make it `true` by default.

kren1 commented on lib/Core/StatsTracker.cpp:
```diff
@@ -285,18 +286,19 @@ void StatsTracker::done() {
+        std::tie(user, sys) = time::getElapsedTime();
+        const auto delta = user - lastUserTime;
+        const auto deltaNow = now - lastNowTime;
+        stats::instructionTime += static_cast<std::uint64_t >(delta);
```
I think it's confusing that this cast silently converts to microseconds. Perhaps `toMicroseconds()`?

kren1 commented on lib/Support/Time.cpp:
```diff
@@ -7,26 +7,197 @@
+  return tv;
+}
+
+time::Span::operator std::uint64_t() const {
```
This is a bit arbitrary?

251 replied:
The cast? It is the underlying type for stats...

kren1 replied:
Yea but reading `static_cast<uint64_t>` does not tell you, you are converting to microseconds ...

kren1 commented on lib/Support/Time.cpp:
```diff
@@ -7,26 +7,197 @@
-  return sys::TimeValue::now();
+
+/// Returns pair (user time, system time) of durations
+std::pair<time::Span, time::Span> time::getElapsedTime() {
```
This name is confusing. Maybe `getElapsedTimeSinceStartup()` ?

kren1 commented on lib/Support/Time.cpp:
```diff
@@ -7,26 +7,197 @@
+ * and not:
+ * - clock_gettime(CLOCK_MONOTONIC_COARSE): software clock, Linux-specific
+ * - clock_gettime(CLOCK_MONOTONIC): slowest on macOS, C-like API
+ * - gettimeofday: C-like API, non-monotonic
```
Maybe add a `TODO` for when we require C++14, to use time literals?

kren1 commented on lib/Support/Time.cpp:
```diff
@@ -7,26 +7,198 @@
+
+  return;
+
+err:
```
Hmmm, I didn't notice this error handling before.  Having constructors fail seem a bit iffy, but I guess since `klee_error` calls `exit` it's fine.

Also `goto`, seriously?

251 commented:
> But I'd like to understand better the change of the option types from integral to string type: is it to allow things like "10ms"? Is the default always seconds?

@ccadar Before we used seconds (mostly), milliseconds (min-query-time-to-log, max-time) and sometimes no unit was specified (e.g. seed-time). Now everything uses seconds (e.g. "10", "0.5", "27.333" but not "1e5") without units or a new format with explicit units, e.g.: "5s", "10min", "3h25min6s" (or even "8h8h"). Supported units are h, min, s, ms, us, ns.

> I feel we need to document this somewhere for users.

I've added a category for the resp. flags in `--help`. Sth. like that?
```
Time-based limits and thresholds:

  Time spans can be specified in two ways:
    1. As positive real numbers representing seconds, e.g. '10', '3.5' but not 'INF', 'NaN', '1e3', '4s'
    2. As a sequence of natural numbers with specified units, e.g. '1h10min' (= '70min'), '5min10s' but not '3.5min', '8S'
       The following units are supported: h, min, s, ms, us, ns.

  -batch-time=<string>                    - Amount of time to batch when using --use-batching-search (default=5s)
  -istats-write-interval=<string>         - Approximate number of seconds between istats writes (default=10s)
  -max-instruction-time=<string>          - Allow a single instruction to take only this much time (default=0s (off)). Enables --use-forked-solver
  -max-solver-time=<string>               - Maximum amount of time for a single SMT query (default=0s (off)). Enables --use-forked-solver
  -max-time=<string>                      - Halt execution after the specified number of seconds (default=0s (off))
  -min-query-time-to-log=<string>         - Set time threshold for queries logged in files. Only queries longer than threshold will be logged. (default=0s)
  -seed-time=<string>                     - Amount of time to dedicate to seeds, before normal search (default=0s (off))
  -stats-write-interval=<string>          - Approximate time between stats writes (default=1s)
  -uncovered-update-interval=<string>     - (default=30s)
```

@MartinNowack In case we keep this category, where should it be defined (file + namespace)?

251 commented:
@ccadar @MartinNowack I played a little bit with CommandLine: it is also possible to

* add a description on top in the OVERVIEW section
* add a flag to multiple categories (categories are sorted by name in the output)

We can also add a man page or update the online documentation but I'm not sure if the online documentation is for the latest release, the master branch or both.

kren1 commented:
I like the `--help` option. A man page is a bit of a large undertaking and online documentation is far away.

251 commented:
I removed the category and added it via `extrahelp` to the end of the help message:
```
Time format used by KLEE's options

  Time spans can be specified in two ways:
    1. As positive real numbers representing seconds, e.g. '10', '3.5' but not 'INF', 'NaN', '1e3', '-4.5s'
    2. As a sequence of natural numbers with specified units, e.g. '1h10min' (= '70min'), '5min10s' but not '3.5min', '8S'
       The following units are supported: h, min, s, ms, us, ns.
```

kren1 commented on include/klee/Internal/System/Time.h:
```diff
@@ -10,31 +10,108 @@
+    /// Returns point in time using a monotonic steady clock
+    Point getWallTime();
+
+    struct Point {
```
I think the `Point` and `Span` class should inherit directly from their `std::chrono` counterparts.

The reasons for not using `std::chrono`s anyway is that they have wrong default initialisation  and annoying names. This can both be solved with inheritance. The other main benefit is that `Point` and `Span` can that interface normally with any library that expects `std::chrono` classes.  As a minor bonus it makes this patch significantly shorter.

It seems like a perfect fit to me, despite being an unconventional solution.

ccadar commented:
@251 can you rebase this, and squash the commit with "address comments"?  I think this would be a good time to merge it.

251 commented:
@ccadar done

ccadar commented:
Thanks again, @251 , really nice patch.
